### PR TITLE
Fix failing grand_canonical sample test and add documentation to samples

### PR DESCRIPTION
**Pull request by jonaslandsgesell**
This PR adds documentation in to the sample files:

* widom_insertion.py
* wang_landau_reaction_ensemble.py
* grand_canonical.py
* reaction_ensemble.py

The PR also fixes partly #3104: The problem with the failing test was that the excess chemical potential did not match the submitted concentration.
I now provide a matching pair of concentration and excess chemical potential.

### CODE REVIEW

File: testsuite/scripts/samples/test_grand_canonical.py [hunk 1]
```diff
@@ -20,7 +20,7 @@
 
 sample, skipIfMissingFeatures = importlib_wrapper.configure_and_import(
     "@SAMPLES_DIR@/grand_canonical.py", n_int_cycles=51, n_int_steps=5,
-    warm_n_times=10, warm_steps=50, cmd_arguments=[0.01, 0.01])
+    warm_n_times=10, warm_steps=50, cmd_arguments=[8.523659461370000200e-04, -3.365753246398797693e-01])
```
Comment (jngrad) [inline testsuite/scripts/samples/test_grand_canonical.py]: how many significant figures do we really need here?

File: samples/wang_landau_reaction_ensemble.py [hunk 2]
```diff
@@ -17,7 +17,9 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #
 """
-This sample simulates the Wang-Landau Reaction Ensemble for a harmonic bond.
+This example script simulates two reacting monomers which are bonded via a harmonic potential.
+The script aborts as soon as the abortion criterion in the Wang-Landau algorithm is met.
+The Wang-Landau potential which is updated during the simulation is written to the file WL_potential_out.dat.
```
Comment (jngrad) [inline samples/wang_landau_reaction_ensemble.py]: Just to check my understanding: we have a neutral monomer (type 3) and a base monomer (type A-) connected by a bond, and two protons (type H+) to set the pH. The WLRE algorithm changes the base monomer into its conjugated acid (type AH) and back using MC moves and stores the degree of association ñ and the chemical potential µ to two files.

File: testsuite/scripts/samples/test_grand_canonical.py (hunk 1 above)
Comment (jonaslandsgesell) [reply testsuite/scripts/samples/test_grand_canonical.py]: 8.523659461370000200e-04 was the concentration I used for determining the excess chemical potential. This is therefore exact. 
The value for the excess chemical potential has a standard error of  +/- 1.590969027075880404e-03 therefore I cropped the result to the third digit after the point: -0.336

File: samples/wang_landau_reaction_ensemble.py (hunk 2 above)
Comment (jonaslandsgesell) [reply samples/wang_landau_reaction_ensemble.py]: The situtation is the following: We have two bonded monomers. These can change their type. They are either of type A- or of type HA. When they are of type A- they have released the corresponding H+ ion. The pH would have to be determined through the reweighted H+ concentration after the simulation is over.
The Wang-Landau algorithm is a histogram reweighting algorithm and drives the system out of equilibrium via applying a Wang-Landau potential. This is not directly connected to the chemical potential. The Wang-Landau potential either represents the canonical partition function, the density of states or the a mixture of both, depending on the acceptance probabilities you impose on the system.
A detailed explaination can be found in https://pubs.acs.org/doi/full/10.1021/acs.jctc.6b00791 .

File: testsuite/scripts/samples/test_grand_canonical.py (hunk 1 above)
Comment (jngrad) [reply testsuite/scripts/samples/test_grand_canonical.py]: you're providing a concentration with 19 significant figures... either WLRE relies on chaos theory, or you're not familiar with Avogadro's constant :) please trim that down, espresso runs on multiple CPU architectures with different machine precision, your value doesn't even fit inside a `long double` on a 64bit machine

File: samples/wang_landau_reaction_ensemble.py (hunk 2 above)
Comment (jngrad) [reply samples/wang_landau_reaction_ensemble.py]: Thanks!

**Review by jonaslandsgesell — COMMENTED** [review]
Looks good to me

**Review by jngrad — APPROVED** [review]

**Pull request merged**